src/utils/dom/createNode.js

```diff
@@ -11,19 +11,27 @@ governing permissions and limitations under the License.
 */
 
 import appendNode from "./appendNode";
+import isObject from "../isObject";
 
-export default (tag, attrs = {}, props = {}, children = [], doc = document) => {
-  const result = doc.createElement(tag);
-
+const populateElementAttributes = (element, attrs) => {
   Object.keys(attrs).forEach(key => {
-    result.setAttribute(key, attrs[key]);
+    if (key === "style" && isObject(attrs[key])) {
+      const styleAttrs = attrs[key];
+      Object.keys(styleAttrs).forEach(styleKey => {
+        element.style[styleKey] = styleAttrs[styleKey];
+      });
+    } else {
+      element.setAttribute(key, attrs[key]);
+    }
```

Comment:
It's there to avoid CSP errors for style attributes. You can set properties on a style attribute but if the string is interpreted CSP will throw an error. If we want to fully support all different ways we set element style attributes we need something like this. The changes to the style in fireReferrerHideableImage.js is an example where if unsafe-inline was removed from style-src and a nonce were added the previsous version of fireReferrerHideableImage.js would result in CSP style-src errors.

Reply:
That's good info, thanks. Considering we have to deal with props for inline styles then, I would prefer we didn't use `populateElementAttributes` at all when dealing with styles. In other words, instead of `fireReferrerHideableImage.js` passing this as an attribute:

```
style: {
  display: "none",
  width: 0,
  height: 0
}
```

I think `fireReferrerHideableImage.js` should pass the style object as a property (as part of the `props` option). Then modify this:

```
Object.keys(props).forEach(key => {
  result[key] = props[key];
});
```

to say if `key` is `style` then apply each style onto `result.style`. That way, we're not conflating attributes and properties.

If we're worried about new code within alloy trying to pass `style` as an attribute, which I don't think we ever want because it would cause more CSP problems (right?), we could throw an error in `populateElementAttributes` if we see a style attribute. I'm on the fence about essentially validating our own code and throwing an error, but I can see this as being an easy bug to introduce and it might be hard to create a blanket functional test now to defend against this kind of bug being introduced later.

In any case, we need a comment somewhere explaining why applying a style attribute is problematic, so that some poor soul doesn't come by later and switch it back.

Also, have you tried rendering personalized content with a style attribute on it? I'm wondering how that plays into this CSP strategy. Can personalization content have script tags in it and how does that work with these changes? cc @ninaceban

Reply:
Yea, switching to handling it in props is probably cleaner. Maybe throwing errors on style if nonce is present would be a valid case. The element style attribute is tricky and that's why I only went half way for support because ideally I think we would also need some server side validation to catch these things before they hit the page.

Reply:
What happens if the customer has a `<style>` or `<script>` element inside their personalized content? Are we adding nonces to them? Should we be? It seems like that would be good.

Reply:
Yes, this is being done with this update as well.

Reply:
@liljenback are you still planning on changing his like we discussed above? Or maybe we miscommunicated?

Reply:
@Aaronius, yea it was changed. It's props now.

Reply:
Sorry, you're right. I misinterpreted code. Thanks!